diff -urN findutils-4.1.7.orig/xargs/xargs.c findutils-4.1.7/xargs/xargs.c
--- findutils-4.1.7.orig/xargs/xargs.c	Mon Feb  4 16:51:29 2002
+++ findutils-4.1.7/xargs/xargs.c	Mon Feb  4 16:53:11 2002
@@ -922,6 +922,6 @@
        [--max-args=max-args] [--no-run-if-empty] [--version] [--help]\n\
        [command [initial-arguments]]\n"),
 	   program_name);
-  fputs (_("\nReport bugs to <bug-findutils@gnu.org>."), stream);
+  fprintf (stream, "%s\n", _("\nReport bugs to <bug-findutils@gnu.org>."));
   exit (status);
 }
