--- findutils-4.1.6/locate/Makefile.in.orig	Sat Oct 14 04:51:46 2000
+++ findutils-4.1.6/locate/Makefile.in	Sat Mar 10 12:38:54 2001
@@ -701,7 +701,7 @@
 	chmod +x $@
 
 install:
-	$(top_srcdir)/mkinstalldirs $(localstatedir)
+	$(top_srcdir)/mkinstalldirs $(DESTDIR)$(localstatedir)
 
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
