--- findutils-4.2.6/configure.in.orig	2004-11-21 13:28:50.000000000 +0100
+++ findutils-4.2.6/configure.in	2004-11-21 23:11:41.238015696 +0100
@@ -100,7 +100,7 @@
 dnl Checks for library functions that are provided by gnulib.
 AC_FUNC_STRFTIME
 AC_REPLACE_FUNCS(memcmp memset stpcpy strdup strstr strtol strtoul)
-AC_CHECK_FUNCS(fchdir getcwd strerror endgrent endpwent)
+AC_CHECK_FUNCS(fchdir getcwd strerror endgrent endpwent setlocale)
 AC_FUNC_VPRINTF
 AC_FUNC_ALLOCA
 AC_FUNC_GETMNTENT
